docs/posts/1tbc/index.qmd

```diff
@@ -0,0 +1,755 @@
+t.order_by(t["l_orderkey"], t["l_partkey"], t["l_suppkey"]).head(3)
+```
+
+![CPU/RAM while Ibis with the DuckDB backend sorting](ibis-duckdb-sort.gif)
+
+## Ibis (DataFusion)
+
+```{python}
+ibis.set_backend("datafusion")
+t = ibis.read_parquet(data)
+t.order_by(t["l_orderkey"], t["l_partkey"], t["l_suppkey"]).head(3)
+```
+
+![CPU/RAM while Ibis with the DataFusion backend sorting](ibis-datafusion-sort.gif)
+
+## Polars (lazy)
+
+```{.python}
+df = pl.scan_parquet(data)
+(
+    df.sort(pl.col("l_orderkey"), pl.col("l_partkey"), pl.col("l_suppkey"))
+    .head(3)
+    .collect()
+)
+```
+
+```html
+The Kernel crashed while executing code in the current cell or a previous cell.
+Please review the code in the cell(s) to identify a possible cause of the failure.
+Click here for more info.
+View Jupyter log for further details.
+```
+
+![CPU/RAM while Polars with the lazy API sorting](polars-lazy-sort.gif)
+
+## Polars (lazy, streaming)
+
+```{.python}
+df = pl.scan_parquet(data)
+(
+    df.sort(pl.col("l_orderkey"), pl.col("l_partkey"), pl.col("l_suppkey"))
+    .head(3)
+    .collect(streaming=True)
+)
+```
+
+```html
+PanicException: called `Result::unwrap()` on an `Err` value: "SendError(..)"
+```
+
+See [GitHub
+issue](https://github.com/pola-rs/polars/issues/17289#issuecomment-2200469528).
+
+![CPU/RAM while Polars with the lazy API, streaming engine sorting](polars-lazy-streaming-sort.gif)
+
+:::
+
+:::{.callout-tip title="Visualize the Ibis expression tree" collapse="true"}
+
+```{python}
+#| code-fold: true
+#| code-summary: "Show code to visualize the Ibis expression tree"
+from ibis.expr.visualize import to_graph
+
+to_graph(t.order_by(t["l_orderkey"], t["l_partkey"], t["l_suppkey"]).head(3))
+```
+
+:::
+
+Ibis with the DuckDB and DataFusion backends complete this in about 2 minutes
+each. Polars (lazy) crashes the kernel after about 2 minutes with its default
+mode and panicks in streaming mode.
+
+**Streaming is an overloaded term here**. In the context of Ibis, a streaming
+backend refers to a near real-time data processing engine like [Apache
+Flink](https://ibis-project.org/backends/flink) or
+[RisingWave](https://ibis-project.org/backends/risingwave). In the context of
+Polars, streaming is a separate engine from the default that can handle
+larger-than-memory data. This general paradigm is already used by DuckDB and
+DataFusion, hence their ability to complete the above query. [The Polars team
+does not recommend using their current streaming engine for
+benchmarking](https://github.com/pola-rs/polars/issues/16694#issuecomment-2146668559)
+and has [announced a new version of their streaming
+engine](https://pola.rs/posts/announcing-polars-1/#new-engine-design).
+
+As we'll see in the benchmark result, some queries will fail to complete with
+Polars and DataFusion. These queries are killed by the operating system due to a
+lack of memory.
+
+:::{.callout-tip title="Sampling large datasets with Ibis" collapse="true"}
```

Comment:
I think this might be missing a closing `:::`

Reply:
it's on line 399, this is a long collapsed example of how to sample w/ Ibis and use pandas or Polars downstream